### PR TITLE
feat(appsync): add L2 constuct for AppSync

### PR DESCRIPTION
**Pull request by hoegertn**
Add L2 constructs for AppSync

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by hoegertn** [comment]
@shivlaks @eladb Any feedback on this? It would be great if we could get this added.

**Comment by shivlaks** [comment]
@hoegertn - hey, thanks for getting the ball rolling. I'll take a look and start getting my thoughts together as we work through these. While I ramp up, are there specific use cases with AppSync we're aiming to start with? Or resources that are challenging to use?

**Comment by hoegertn** [comment]
Hi, I started with datasources of type DynamoDB and Lambda, but I can provide the rest if the architecture is ok. 

The main features/benefits are:
* Support for higher-level constructs Table, Function, etc
* Auto-creation of IAM Roles and Policies
* File access to extract schema, mapping templates, etc

File: packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+    private resolver: CfnResolver;
+
+    constructor(scope: Construct, id: string, props: ResolverProps) {
+        super(scope, id);
+
+        this.resolver = new CfnResolver(this, 'Resource', {
+            apiId: props.api.apiId,
+            typeName: props.typeName,
+            fieldName: props.fieldName,
+            dataSourceName: props.dataSource.name,
```
Comment (ranguard) [inline packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts]: Small spot... `dataSourceName` is optional - you can have resolvers with no data source

Comment (hoegertn) [reply packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts]: I changed it, thanks.

**Comment by hoegertn** [comment]
@shivlaks Do you think we can proceed here? I would love to have this in a release any time soon, as I am using this for several projects already.

File: packages/@aws-cdk/aws-appsync/README.md [hunk 2]
```diff
@@ -16,3 +16,94 @@
...
+    const customerTable = new Table(this, 'CustomerTable', {
+      billingMode: BillingMode.PAY_PER_REQUEST,
+      partitionKey: {
+        name: 'id',
+        type: AttributeType.STRING,
+      },
+    });
+    const customerDS = api.addDynamoDbDataSource('Customer', 'The customer data source', customerTable);
+    customerDS.createResolver({
+      typeName: 'Query',
```
[the remark this answers is not in the document]
Comment (hoegertn) [reply packages/@aws-cdk/aws-appsync/README.md]: THis is a free text and could be anything the user wants

File: packages/@aws-cdk/aws-appsync/README.md [hunk 3]
```diff
@@ -16,3 +16,94 @@
...
+      fieldName: 'getCustomers',
+      requestMappingTemplate: MappingTemplate.dynamoDbScanTable(),
+      responseMappingTemplate: MappingTemplate.dynamoDbResultList(),
+    });
+    customerDS.createResolver({
+      typeName: 'Query',
+      fieldName: 'getCustomer',
+      requestMappingTemplate: MappingTemplate.dynamoDbGetItem('id', 'id'),
+      responseMappingTemplate: MappingTemplate.dynamoDbResultItem(),
```
[the remark this answers is not in the document]
Comment (hoegertn) [reply packages/@aws-cdk/aws-appsync/README.md]: You can also write `MappingTemplate.fromFile()` or `MappingTemplate.fromString()`

File: packages/@aws-cdk/aws-appsync/test/integ.graphql.expected.json [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+                "dynamodb:BatchGetItem",
+                "dynamodb:GetRecords",
+                "dynamodb:GetShardIterator",
+                "dynamodb:Query",
+                "dynamodb:GetItem",
+                "dynamodb:Scan",
+                "dynamodb:BatchWriteItem",
+                "dynamodb:PutItem",
+                "dynamodb:UpdateItem",
+                "dynamodb:DeleteItem"
```
[the remark this answers is not in the document]
Comment (hoegertn) [reply packages/@aws-cdk/aws-appsync/test/integ.graphql.expected.json]: It is using 
```
	if (props.readOnlyAccess) {
        props.table.grantReadData(this.serviceRole);
    } else {
        props.table.grantReadWriteData(this.serviceRole);
    }
```

File: packages/@aws-cdk/aws-appsync/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (shivlaks) [reply packages/@aws-cdk/aws-appsync/README.md]: good point, finally started reading the [user guide](https://docs.aws.amazon.com/appsync/latest/devguide/appsync-dg%20.pdf) to get caught up here

**Comment by hoegertn** [comment]
IF you do not have major changes, I could merge master in later today and we could get this merged as a first version and add more data sources in further PRs when needed. This construct is working, as I am using it in several prod setups now.

**Comment by hoegertn** [comment]
@shivlaks can we get this merged?

**Comment by shivlaks** [comment]
> @shivlaks can we get this merged?

Hey @hoegertn - code looks fine to me! Seems like there's still a conflict that needs resolution.

**Comment by hoegertn** [comment]
I merged in the latest release. It would be great if we could get this into master.

**Comment by shivlaks** [comment]
@hoegertn LTGM. hang tight, we'll start merging community contributions again in a couple days!

**Comment by hoegertn** [comment]
@shivlaks Hi, can you please approve again? I had to merge master in, due to conflicts.

**Comment by shivlaks** [comment]
@hoegertn - done, excited to start growing this L2 and filling in the rest!

**Pull request merged**